File: src/main/java/net/igneo/icv/entity/wave/WaveEntity.java
package net.igneo.icv.entity.wave;

import net.igneo.icv.entity.ICVEntity;
import net.igneo.icv.entity.surfWave.SurfWaveEntity;
import net.igneo.icv.init.ICVUtils;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.projectile.Projectile;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;

public class WaveEntity extends ICVEntity {
    public WaveEntity(EntityType<? extends Projectile> pEntityType, Level pLevel) {
        super(pEntityType, pLevel);
    }
    private int lifetime = 0;
    private Vec3 storedMotion = Vec3.ZERO;
    public void setTrajectory(Vec3 vec) {
        storedMotion = vec;
    }

    @Override
    public float getStepHeight() {
        return 1.5F;
    }

    @Override
    public void tick() {
        super.tick();
        for (Entity entity : this.level().getEntities(null,this.getBoundingBox().inflate(3))) {
            if (!(entity instanceof WaveEntity) && !(entity instanceof SurfWaveEntity)) {
                entity.addDeltaMovement(storedMotion);
            }
        }
        setDeltaMovement(storedMotion);
        if (lifetime < 20) {
            ++lifetime;
        } else {
            this.discard();
        }
    }

    @Override
    protected boolean canAddPassenger(Entity pPassenger) {
        return pPassenger == this.getOwner();
    }
}

File: src/main/java/net/igneo/icv/entity/leggings/wave/WaveEntity.kt
package net.igneo.icv.entity.leggings.wave

import net.igneo.icv.entity.ICVEntity
import net.igneo.icv.entity.boots.surfWave.SurfWaveEntity
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.projectile.Projectile
import net.minecraft.world.level.Level
import net.minecraft.world.phys.Vec3

class WaveEntity(pEntityType: EntityType<out Projectile?>, pLevel: Level) :
    ICVEntity(pEntityType, pLevel) {
    private var lifetime = 0
    private var storedMotion: Vec3 = Vec3.ZERO

    fun setTrajectory(vec: Vec3) {
        storedMotion = vec
    }

    override fun getStepHeight(): Float {
        return 1.5f
    }

    override fun tick() {
        super.tick()
        for (entity in level().getEntities(null, this.boundingBox.inflate(3.0))) {
            if (entity !is WaveEntity && entity !is SurfWaveEntity) {
                entity.addDeltaMovement(storedMotion)
            }
        }
        deltaMovement = storedMotion
        if (lifetime < 20) {
            ++lifetime
        } else {
            this.discard()
        }
    }

    override fun canAddPassenger(pPassenger: Entity): Boolean {
        return pPassenger === this.owner
    }
}
